### ISSUE TITLE
OSX keychain entry disappearing for yubikey smartcard

### ISSUE THREAD

**Issue by huang0320** [OpenSC/OpenSC#1479]
### Problem Description

1) using Yubikey PIV
2) getting certificate on the slot 9a and 9d
3) keychain see PIV entry for sometime, then disappear (have to pull/reinsert key to get it to show)

### Proposed Resolution
just like to know the correct path to make sure keychain entries for PIV stay consistent.   this is specific to OSX sierra and high sierra release with both openSC (.17 and .18) and Yubikey 4

### Steps to reproduce
see above

**Comment by frankmorgner** [comment]
@mouse07410 ?

**Comment by mouse07410** [comment]
I observe this (inability to detect the inserted token itself) very rarely. I see **no correlation** with what certificate slot is used (9a, 9c, 9d, etc.).

Much more frequent in my experience is having the token "stuck", aka the system fails to notice that the token was removed. It would still prompt me for a PIN for privileged operations, display the token in the Keychain Access, etc. The only remedy I found was re-inserting the token and removing it after a few seconds.

I think that both my and the OP's problems seem to relate to the same cause: issues with detecting change in the token/slot status (inserted, or removed).

This happened on Sierra, and continues to High Sierra. It is an annoyance, but not a huge one...


> just like to know the correct path to make sure keychain entries for PIV stay consistent

I don't understand what this means, or what path you're talking about. Keychain entries for PIV in my experience seem consistent enough, it's the presence or absence of the token that isn't always detected/propagated correctly.

**Comment by huang0320** [comment]
thanks for the quick response.   in my case, the keychain folder for PIV exists.   However, the folder may be empty after  I clicked on it (even the key is in the slot) and I have to remove/reinsert the key to get it shown up again.   I also experienced the stuck case.   I'm more concerns about the PIV keychain folder does not always get created (especially with touchID equipped Mac).  Is there any inside to that?
